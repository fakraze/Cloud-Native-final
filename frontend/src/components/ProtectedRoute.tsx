import React from 'react';
import { Navigate } from 'react-router-dom';
import { useAuthStore } from '../store/authStore';
import { User } from '../types/auth';

interface ProtectedRouteProps {
  children: React.ReactNode;
  requiredRole?: User['role'];
}

const ProtectedRoute: React.FC<ProtectedRouteProps> = ({ children, requiredRole }) => {
  const { isAuthenticated, user } = useAuthStore();

  if (!isAuthenticated) {
    return <Navigate to="/login" replace />;
  }

  if (requiredRole && user?.role !== requiredRole) {
    return <Navigate to="/restaurant" replace />;
  }
  return <>{children}</>;
};

export { ProtectedRoute };
