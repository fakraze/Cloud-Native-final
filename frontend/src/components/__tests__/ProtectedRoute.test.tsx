import { describe, it, expect, beforeEach, vi } from 'vitest';
import { render } from '../../test/test-utils';
import { ProtectedRoute } from '../ProtectedRoute';

const mockAuthStore = vi.hoisted(() => ({
  useAuthStore: vi.fn()
}));

vi.mock('../../store/authStore', () => mockAuthStore);

describe('ProtectedRoute Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('should render children when user is authenticated and has access', () => {
    mockAuthStore.useAuthStore.mockReturnValue({
      isAuthenticated: true,
      user: {
        id: '1',
        username: 'testuser',
        role: 'customer'
      }
    });

    const { container } = render(
      <ProtectedRoute>
        <div>Protected Content</div>
      </ProtectedRoute>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with role restrictions', () => {
    mockAuthStore.useAuthStore.mockReturnValue({
      isAuthenticated: true,
      user: {
        id: '1',
        username: 'admin',
        role: 'admin'
      }
    });

    const { container } = render(
      <ProtectedRoute allowedRoles={['admin']}>
        <div>Admin Only Content</div>
      </ProtectedRoute>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should handle unauthenticated user', () => {
    mockAuthStore.useAuthStore.mockReturnValue({
      isAuthenticated: false,
      user: null
    });

    const { container } = render(
      <ProtectedRoute>
        <div>Protected Content</div>
      </ProtectedRoute>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
