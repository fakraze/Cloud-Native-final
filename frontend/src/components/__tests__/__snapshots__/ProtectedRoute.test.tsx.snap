// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProtectedRoute Component > should handle unauthenticated user 1`] = `null`;

exports[`ProtectedRoute Component > should render children when user is authenticated and has access 1`] = `
<div>
  Protected Content
</div>
`;

exports[`ProtectedRoute Component > should render with role restrictions 1`] = `
<div>
  Admin Only Content
</div>
`;
